File: Chapter 1/1030.py
k = int(input())
rice = []

for i in range(k):
    p,q = [float(x) for x in input().split()]
    rice.append([p/q,q]) # เก็บ[ราคาต่อ กก , จำนวน กก ที่มี]
rice.sort()
m = int(input())
i = 0
price = 0
j = 0
while i < m:
    b = int(input())
    if b >= rice[j][1]:
        price += rice[j][1]*rice[j][0]
        j += 1
        i += 1
    else:
        rice[j][1] -= b
        price += rice[j][0]*b
        i += 1      
    print("{:.3f}".format(price))
        

File: C plus/1030.cpp
#include <bits/stdc++.h>
using namespace std;
struct Rice{
    double myP;
    double myQ;
};
int compare(Rice r1,Rice r2){
    return r1.myP < r2.myP;
}
int main(){
    std::cout << std::setprecision(3) << std::fixed;
    int k,m,b;
    double p,q,price;
    cin >> k;
    Rice rice[k];
    for (int i = 0;i<k;i++){
        cin >> p >> q;
        rice[i].myP = p/q;
        rice[i].myQ = q;
    }
    sort(rice,rice+k,compare);
    cin >> m;
    for (int i = 0;i<m;i++){
        cin >> b;
        price = 0;
        int start = 0;
        double P,Q;
        while(b>0){
            P = rice[start].myP;
            Q = rice[start].myQ;
            if (b >= Q){
                b -= Q;
                price += (P)*(Q);
                rice[start].myQ = 0;
                start++;
            }else{
                rice[start].myQ -= b;
                price += (P)*b;
                b = 0;
            }
        }
    cout << price << endl;
    }
    return 0;
}